src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java

```diff
@@ -48,74 +54,159 @@
+          Map<Digest, List<Path>> toUpload = new HashMap<>();
+
+          // mark all entries without a PathDigestPair for upload
+          for (Pair<Path, PathDigestPair> pathPair : pairs) {
+            Path path = pathPair.first;
+            if (pathPair.second == null) {
+              DigestUtil digestUtil = new DigestUtil(path.getFileSystem().getDigestFunction());
+              Digest digest = digestUtil.compute(path);
+              if (!toUpload.containsKey(digest)) {
+                toUpload.put(digest, new LinkedList<>());
+              }
+              toUpload.get(digest).add(path);
+            } else {
+              pathDigestFutures.add(Futures.immediateFuture(pathPair.second));
+            }
+          }
+
+          Context prevContext = ctx.attach();
+          ListenableFuture<ImmutableSet<Digest>> missingDigestsFuture =
+              digestLookup.getMissingDigests(toUpload.keySet());
+          ctx.detach(prevContext);
+
+          // fetch missing digests and upload those files that aren't in the CAS
+          ListenableFuture<List<PathDigestPair>> uploadedPathDigestFutures = Futures.transformAsync(
+              missingDigestsFuture,
+              missingDigests -> {
+                // new hashset because removeAll mutates the dictionary
+                Set<Digest> cachedDigests = new HashSet<>(toUpload.keySet());
+                cachedDigests.removeAll(missingDigests);
+
+                List<ListenableFuture<PathDigestPair>> cachedPairFutures = new ArrayList<>();
+
+                // if not on the CAS, add the result of the upload
+                for (Digest digest : missingDigests) {
+                  for (Path path : toUpload.get(digest)) {
+                    cachedPairFutures.add(uploadFile(path, digest));
+                  }
                 }
-                Chunker chunker = Chunker.builder().setInput(digest.getSizeBytes(), file).build();
-                final ListenableFuture<Void> upload;
-                Context prevCtx = ctx.attach();
-                try {
-                  upload =
-                      uploader.uploadBlobAsync(
-                          HashCode.fromString(digest.getHash()), chunker, /* forceUpload=*/ false);
-                } finally {
-                  ctx.detach(prevCtx);
+
+                // otherwise, if cached, just add the file
+                for (Digest digest : cachedDigests) {
+                  for (Path path : toUpload.get(digest)) {
+                    cachedPairFutures.add(Futures.immediateFuture(new PathDigestPair(path, digest)));
+                  }
                 }
-                return Futures.transform(
-                    upload, unused -> new PathDigestPair(file, digest), uploadExecutor);
+
+                return Futures.allAsList(cachedPairFutures);
               },
-              MoreExecutors.directExecutor());
-      uploads.add(digestFuture);
-      }
+              MoreExecutors.directExecutor()
+          );
```

Comment:
I think the code would be easier to read if the lambdas passed to `Futures.transformAsync` would be methods.

Reply:
Will do.